Remove legacy Jibri config.json.

saygun opened the pull request:
Hello everyone.

As mentioned in this issue #890, I replaced the legacy configuration file with the new one (`jibri.conf`)

I also added a few new `ENV` variables which we are already using in our Jibri setup. 

* `CHROMIUM_FLAGS`: comma seperated chromium flags.
ex: `--disable-setuid-sandbox,--user-agent=some_custom_user_agent,--ignore-certificate-errors`
* `XMPP_TRUST_ALL_CERTS`: defaults to `false`
* `JIBRI_USAGE_TIMEOUT`:  defaults to `0`
* `JIBRI_HTTP_API_EXTERNAL_PORT`: defaults to `2222`
* `JIBRI_HTTP_API_INTERNAL_PORT`: defaults to `3333`
* `ENABLE_STATS_D`: defaults to `false`

Please share your concerns. @saghul @sapkra @netaskd @aaronkvanmeerten

Buyomer commented:
👍

prayagsingh commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,58 @@
+jibri {
+    api {
+        http {
+            external-api-port = {{ .Env.JIBRI_HTTP_API_EXTERNAL_PORT | default 2222 }}
```
Instead of using `default`(hardcoded value), I would recommend using env variable in such a way that if it's not configured then it should be referenced from `reference.conf` file.  @bgrozev recommended this approach when I created new jicofo config PR. 
for example. 
```
{{  if .Env.JIBRI_HTTP_API_EXTERNAL_PORT }}
external-api-port = {{ .Env.JIBRI_HTTP_API_EXTERNAL_PORT }}
{{ end }}

```

bbaldino replied:
Yes, this is the preferred way: the default value is already encoded in `reference.conf` in the jibri repo, so we don't want another place where a default is set.

bbaldino commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,58 @@
+    api {
+        http {
+            external-api-port = {{ .Env.JIBRI_HTTP_API_EXTERNAL_PORT | default 2222 }}
+            internal-api-port = {{ .Env.JIBRI_HTTP_API_INTERNAL_PORT | default 3333 }}
```
Same as above

bbaldino commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,58 @@
+        }
+        xmpp {
+            environments = [
+   		        {
```
nit: indentation of this opening brace is off.

bbaldino commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,58 @@
+        xmpp {
+            environments = [
+   		        {
+                    name = "prod environment"
```
nit: Do other dockerfiles use strings like this as a default?  I think we may want to make this pull from a variable.

bbaldino commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,58 @@
+        {{ end -}}
+    }
+    stats {
+        enable-stats-d = {{ .Env.ENABLE_STATS_D | default false }}
```
No defaults here

prayagsingh commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,58 @@
+   		        {
+                    name = "prod environment"
+                    xmpp-server-hosts = [
+                        "{{ .Env.XMPP_SERVER }}"
```
Allow to add multiple XMPP servers. We can configure `XMPP_SERVER =xmpp1,xmpp2` using env variables. You can use something like this `"{{ join "\",\n\"" (splitList "," .Env.XMPP_SERVER }}"`

prayagsingh commented:
@saygun @bbaldino I made some changes. 
```
{{ $ENABLE_JIBRI_HTTP_API_PORT := .Env.ENABLE_JIBRI_HTTP_API_PORT | default "1" | toBool }}
jibri {
    // A unique identifier for this Jibri
    // TODO: eventually this will be required with no default
    id = "jibri-instance-{{ .Env.JIBRI_INSTANCE_ID }}"

    api {
      {{ if $ENABLE_JIBRI_HTTP_API_PORT }}  
      http {
        {{  if .Env.JIBRI_HTTP_API_EXTERNAL_PORT }}
        external-api-port = {{ .Env.JIBRI_HTTP_API_EXTERNAL_PORT }}
        {{ end }}
        {{ if .Env.JIBRI_HTTP_API_INTERNAL_PORT }}
        internal-api-port = {{ .Env.JIBRI_HTTP_API_INTERNAL_PORT }}
        {{ end }}
      }
      {{ end }}
      xmpp {
        // See example_xmpp_envs.conf for an example of what is expected here
        environments = [
            {
                // A user-friendly name for this environment
                name = "prod environment"
    
                // A list of XMPP server hosts to which we'll connect
                xmpp-server-hosts = [ 
                    "{{ .Env.XMPP_SERVER }}"
                ]
    
                // The base XMPP domain
                xmpp-domain = "{{ .Env.XMPP_DOMAIN }}"
    
                // The MUC we'll join to announce our presence for
                // recording and streaming services
                control-muc {
                    domain = "{{ .Env.XMPP_INTERNAL_MUC_DOMAIN }}"
                    room-name = "{{ .Env.JIBRI_BREWERY_MUC }}"
                    nickname = "jibri-instance-{{ .Env.JIBRI_INSTANCE_ID }}"
                }
    
                // The login information for the control MUC
                control-login {
                    domain = "{{ .Env.XMPP_AUTH_DOMAIN }}"
                    username = "{{ .Env.JIBRI_XMPP_USER }}"
                    password = "{{ .Env.JIBRI_XMPP_PASSWORD }}"
                }
                
                sip-control-muc {
                    domain = "{{ .Env.XMPP_INTERNAL_MUC_DOMAIN }}"
                    room-name = "{{ .Env.JIBRI_BREWERY_MUC }}"
                    nickname = "jibri-instanse-{{ .Env.JIBRI_INSTANCE_ID }}"
                }

                // The login information the selenium web client will use
                call-login {
                    domain = "{{ .Env.XMPP_RECORDER_DOMAIN }}"
                    username = "{{ .Env.JIBRI_RECORDER_USER }}"
                    password = "{{ .Env.JIBRI_RECORDER_PASSWORD }}"
                }
                
                // The value we'll strip from the room JID domain to derive
                // the call URL
                strip-from-room-domain = "{{ .Env.JIBRI_STRIP_DOMAIN_JID }}."
    
                {{ if .Env.JIBRI_USAGE_TIMEOUT -}}
                // How long Jibri sessions will be allowed to last before
                // they are stopped.  A value of 0 allows them to go on
                // indefinitely
                usage-timeout = "{{ .Env.JIBRI_USAGE_TIMEOUT }}"
                {{ end -}}    
                
                {{ if .Env.XMPP_TRUST_ALL_CERTS -}}
                // Whether or not we'll automatically trust any cert on
                // this XMPP domain
                trust-all-xmpp-certs = {{ .Env.XMPP_TRUST_ALL_CERTS }}
                {{ end -}}
            }
        ]
      }
    }
    recording {
      recordings-directory = "{{ .Env.JIBRI_RECORDING_DIR }}"
      finalize-script = "{{ .Env.JIBRI_FINALIZE_RECORDING_SCRIPT_PATH }}"
    }
    {{ if .Env.CHROMIUM_FLAGS -}}
    chrome {
      // The flags which will be passed to chromium when launching
      flags = [
        "{{ join "\",\"" (splitList "," .Env.CHROMIUM_FLAGS) }}"
      ]
    }
    {{ end -}}

    {{ if .Env.ENABLE_STATS_D }}
    stats {
      enable-stats-d = {{ .Env.ENABLE_STATS_D }}
    }
    {{ end -}}
}
```

saygun commented:
@prayagsingh It's terrific, thank you so much 🚀 . I will check the comments and update the pull request accordingly today.

saygun commented:
Thank you for your feedback guys. @prayagsingh @bbaldino. I just added another commit. You are right about putting default values. Removed them and encapsulated those optional blocks with if statements as @prayagsingh suggested.

prayagsingh commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,100 @@
+    id = "jibri-instance-{{ .Env.JIBRI_INSTANCE_ID }}"
+    
+    api {
+        {{ if $ENABLE_JIBRI_HTTP_API_PORT -}}
+        http {
+            {{  if .Env.JIBRI_HTTP_API_EXTERNAL_PORT -}}
+            external-api-port = {{ .Env.JIBRI_HTTP_API_EXTERNAL_PORT }}
+            {{ end -}}
+            {{ if .Env.JIBRI_HTTP_API_INTERNAL_PORT -}}
+            internal-api-port = {{ .Env.JIBRI_HTTP_API_INTERNAL_PORT }}
+            {{ end -}}
```
this is more clean approach. If any of the value is not configured then this will fallback to reference.conf.
```suggestion
      {{ if and .Env.JIBRI_HTTP_API_EXTERNAL_PORT .Env.JIBRI_HTTP_API_INTERNAL_PORT }}
      http {
        external-api-port = {{ .Env.JIBRI_HTTP_API_EXTERNAL_PORT }}       
        internal-api-port = {{ .Env.JIBRI_HTTP_API_INTERNAL_PORT }}
      }
      {{ end }}
```

saygun replied:
What if the user want's to configure only `external-api-port` and not the `internal-api-port`.  If I am not mistaken, `and` logic doesn't provide this feature here. What do you think?

prayagsingh replied:
yes it doen't. 

Since there is no enable/disable here hence even if you set `ENABLE_JIBRI_HTTP_API_PORT =0` then it'll be served from reference.conf hence there is no purpose of using this env variable. 

we have to set the default values of both of the env variables i.e `JIBRI_HTTP_API_EXTERNAL_PORT=3333` and `JIBRI_HTTP_API_INTERNAL_PORT=2222` in env.example/.env files. Lets say you changed `JIBRI_HTTP_API_EXTERNAL_PORT =4444` but didn't change `JIBRI_HTTP_API_EXTERNAL_PORT ` then  `JIBRI_HTTP_API_INTERNAL_PORT ` will be served from `env.example/.env` file.

prayagsingh replied:
@saghul Can you please suggest which approach is better here?

bbaldino replied:
I'm tempted to agree with what @saygun's point about having the flexibility to override only a single one of the ports, but @prayagsingh I'm not sure I followed what you were saying there.

saghul replied:
Not sure I followed either. I think they should be settable independently and you can use `| default "0" | toBool` to check if it got disabled by setting it to empty or 0, for example.

prayagsingh replied:
@saghul @bbaldino  I'm trying to avoid an extra env variable. I have updated my suggestion. Please check.

saghul replied:
Looks good!

prayagsingh commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,100 @@
+{{ $ENABLE_JIBRI_HTTP_API_PORT := .Env.ENABLE_JIBRI_HTTP_API_PORT | default "1" | toBool }}
```
Not required anymore. 
```suggestion
```

saghul commented on jibri/rootfs/defaults/jibri.conf:
```diff
@@ -0,0 +1,100 @@
+                        password = "{{ .Env.JIBRI_XMPP_PASSWORD }}"
+                    }
+
+                    sip-control-muc {
```
This is for SIP-Jibri right? Let's drop it for now please.

saygun replied:
Done it 👍

saygun commented:
@prayagsingh @saghul Thank you for your guides. I updated the pull request. Please take a look at it whenever you are available.

saygun commented:
Hello everyone. @saghul @bbaldino Would you mind if I ask you to review this pull request?

bbaldino reviewed (approved):
From what I can tell it looks good...I assume testing with it is working fine?  Thanks for your work on this!

prayagsingh commented:
@bbaldino yes it is working fine. I'm using it from last week. @saygun Great Work 👍

saygun commented:
> @bbaldino yes it is working fine. I'm using it since last week. @saygun Great Work 👍

Thank you 👍 . Credits go to you

saghul commented:
Is this config working with latest stable too?

prayagsingh commented:
> 
> 
> Is this config working with latest stable too?

Didn't tried it stable. I'm building my docker images from unstable all the time.

saygun commented:
> Is this config working with latest stable too?

Yes. I have tested it.

saghul commented:
Thanks folks, I'll take another look soon!

prayagsingh commented:
Hello @saygun Can you please add the FFmpeg config too? It's really helpful when using Pulseaudio. Since I'm trying to use Pulseaudio therefore I have tested it too. 
reference: https://github.com/jitsi/jibri/blob/26eeb53bd26710df78e722e339228f857b8bcc20/src/main/resources/reference.conf#L34

**Snippet of how I configured it in jibri.conf**
```
ffmpeg {
      {{ if .Env.FFMPEG_RESOLUTION -}}
      resolution =  "{{ .Env.FFMPEG_RESOLUTION }}" #"1920x1080"
      {{ end -}}
      {{ if .Env.FFMPEG_AUDIO_SOURCE -}}
      // The audio source that will be used to capture audio on Linux
      audio-source = "{{ .Env.FFMPEG_AUDIO_SOURCE }}"  #"alsa"
      {{ end -}}
      {{ if .Env.FFMPEG_AUDIO_DEVICE -}}
      // The audio device that will be used to capture audio on Linux
      audio-device =  "{{ .Env.FFMPEG_AUDIO_DEVICE }}" #"plug:bsnoop"
      {{ end -}}
    }
```

bbaldino commented:
@prayagsingh that looks good for ffmpeg config, but let's not include the defaults in there (e.g. `#"1920x1080"`)

prayagsingh commented:
@saghul I think it's good to go. Please add the `ffmpeg` part too. I have updated the above ffmpeg config env variables with `JIBRI` as a prefix. Please let me know if you want to test it with pulse-audio as well.

saghul commented:
The ffmpeg options require unstable don't they?

saghul commented:
I'd be happy to switch to PulseAudio since it should simplify the setup!

prayagsingh commented:
> 
> 
> The ffmpeg options require unstable don't they?

Yes it requires unstable version.

prayagsingh commented:
> 
> 
> I'd be happy to switch to PulseAudio since it should simplify the setup!

I have tested it with the below setups.

Image: `prayagsingh/jibri:pulse_v2.1`

**Setup 1: docker-compose/swarm setups** WORKING
 I have tested it using docker-compose and docker-swarm setup and working fine on both the setups( didn't cross 4vCPU and 4Gb memory limits with 3 participants). 

**Setup 2: K8s** Not WORKING
Now I'm trying to use it on K8s setup(ultimate goal) and I don't know why it isn't working. When I start the recording suddenly the memory consumption went up to 7.X Gb and vCPU to 3.86 (with 2 users and screen sharing) and within 3 minutes it came back under 4Gb Memory and 4 vCPU limits. Everything went well for approx 15 minutes. Then the third participant joined and everything was fine for the next 5 minutes. finally, a sudden spike in memory and it crashed.  

**NOTE:** It is mandatory to map `/dev/shm` even if using `--no-sandbox,–disable-dev-shm-usage` flags with chrome else chrome crashes when doing screen-share. 

![Screenshot (617)](https://user-images.githubusercontent.com/8455114/114269850-bf121480-9a26-11eb-861c-b120262936c2.png)


**Env Variables Config**
```
            # new config
            ENABLE_JIBRI_HTTP_API_PORT: 1
            JIBRI_HTTP_API_EXTERNAL_PORT: 2222
            JIBRI_HTTP_API_INTERNAL_PORT: 3333
            JIBRI_USAGE_TIMEOUT: 0
            XMPP_TRUST_ALL_CERTS: "true"
            CHROMIUM_FLAGS: "--use-fake-ui-for-media-stream,--start-maximized,--kiosk,--enabled,--disable-infobars,--autoplay-policy=no-user-gesture-required"
            ENABLE_STATS_D: "true"
            JIBRI_FFMPEG_RESOLUTION: "1920x1080"  #"1280x720"
            JIBRI_FFMPEG_AUDIO_SOURCE: "pulse"
            JIBRI_FFMPEG_AUDIO_DEVICE: "default"
```

**NOTE 2: You can stream to any rtmp server and to facebook too. plus support for rclone is present too**

saghul commented:
This needs a rebase, sorry.
